### ISSUE TITLE
Z3 segfault in versions 4.7.0, 4.7.1, 4.8.2

### ISSUE THREAD

**Issue by invd** [Z3Prover/z3#1914]
During my work with the [Ultimate program analysis framework](https://github.com/ultimate-pa/ultimate) - which uses Z3 extensively - I have tripped over several segfaults that appear to come from one consistent bug.

After first [encountering](https://github.com/ultimate-pa/ultimate/issues/321#issuecomment-403048854) the issue in July, I've tried to localize the error (or at least narrow it down) and extract the corresponding SMT input that provokes the crash before reporting it here, but have so far been unable to do so, as described in the related [Ultimate issue](https://github.com/ultimate-pa/ultimate/issues/376).
I'm aware that the lack of a crashing input file complicates any debugging efforts, but I have so far been unable to reproduce the segfault independently of Ultimate runs, which in this case take several hours to reach the crash or have other issues, which complicates the debugging.

**4.7.0**:
`z3[15276]: segfault at 0 ip 0000000000cf90f2 sp 00007ffe8537c9c0 error 4 in z3[400000+15c5000]`

After swapping the 4.7.0 Z3 binary with the official **4.7.1** binary ([info](https://github.com/ultimate-pa/ultimate/issues/321#issuecomment-404154890)), the problem persists:
`segfault at 0 ip 0000000000cb0f9f sp 00007ffe390b8b90 error 4 in z3[400000+130a000]`

For post-4.7.1-versions, we first thought that the underlying issue has been solved, but the segfaults were seen again with **4.8.2**:
`segfault at 402fcb9ff ip 0000000000bdc983 sp 00007ffe9332fde0 error 4 in z3[400000+15aa000]`

Z3 is called with either `z3 -smt2 SMTLIB2_COMPLIANT=true -in` or explicit limits `z3  -smt2 SMTLIB2_COMPLIANT=true -in -memory:2024 -t:2000` and receives the data on STDIN.

**Comment by invd** [comment]
The logs confirms that the affected Z3 instances are created with `z3 SMTLIB2_COMPLIANT=true -memory:2024 -smt2 -in -t:2000`. As written before, we've previously tried to extract all smt scripts that are passed to Z3 in this situation and have been unable to reproduce the crash this way. @danieldietsch will try this again shortly.

Unfortunately, the current situation makes attaching gdb difficult and the Z3 build we're using contains no debug symbols. For this reason, I've previously compiled v4.7.1 with the "RelWithDebugInfo" target.
`segfault at 0 ip 000055e25fe2d282 sp 00007ffcb23b3670 error 4 in z3_v4-7-1_relwithdebuginfo[55e25f5c8000+1323000]`
In this case, the error appears to be located at offset `0x865282` in an object of size `0x1323000`.

Error 4 translates to "The cause was a user-mode read resulting in no page being found."

addr2line gives the following error location:
```
addr2line -e z3_v4-7-1_relwithdebuginfo -fCi 0x865282                                                                                                                     
bit_blaster_tpl<bit_blaster_cfg>::mk_neg(unsigned int, expr* const*, ref_vector<expr, ast_manager>&)
/home/masterarbeit/z3rel/src/ast/rewriter/bit_blaster/bit_blaster_tpl_def.h:113
```
This points to https://github.com/Z3Prover/z3/blob/520ce9a5ee6079651580b6d83bc2db0f342b8a20/src/ast/rewriter/bit_blaster/bit_blaster_tpl_def.h#L113

Does this error location look reasonable for the encountered error?

**Comment by NikolajBjorner** [comment]
This is a start, but a fuller stack really helps. Are you indicating that you can't break in a debugger when there is an error? 
The error indicates that the size passed in does not correspond to the space allocated by the vector.

**Comment by danieldietsch** [comment]
I can reproduce the segfault with the script [partialmain_bmp_logo64_evil_Elim1Store.c.zip](https://github.com/Z3Prover/z3/files/2550495/partialmain_bmp_logo64_evil_Elim1Store.c.zip). Its about 95MB when extracted. 

I use ``z3 SMTLIB2_COMPLIANT=true -memory:2024 -smt2 -t:2000`` on a z3 compiled from 5d06fa234 (z3-4.8.1-master-5d06fa234 from 2018-10-29). 

We compile with ``--staticbin``, perhaps this might also influence the issue.

**Comment by NikolajBjorner** [comment]
thanks, this is useful.

**Comment by invd** [comment]
My self-compiled `z3_v4-7-1_relwithdebuginfo` **does not segfault** on this input (!), so this error behavior has likely shifted between these versions.

**Comment by NikolajBjorner** [comment]
if it depends on timeout behavior, would be useful to run with different random seeds and timeouts

**Comment by invd** [comment]
`z3_relWithDebug_4-8-0` does segfault:
```
z3_relWithDebug[18664]: segfault at 559a723511ff ip 000055966fb852c3 sp 00007ffc2343d6f0 error 4 in z3_relWithDebug_4-8-0[55966f1b2000+12e1000]
```
Time for gdb ..

**Comment by NikolajBjorner** [comment]
the first stack is the same I encountered and the push fixes this one.
I ran again in both debug and release mode on the fixed version and didn't hit bugs yet.
Trying with different random seeds and timeouts, but each run is pretty long.
I am unsure about the valgrind messages. The parameters is very stable code and has been through other analyzers.

**Comment by invd** [comment]
Valgrind has run through (after a pretty long run due to the overhead) and shows a similar backtrace:
```
==24056== Process terminating with default action of signal 11 (SIGSEGV)
==24056==  Access not within mapped region at address 0x4044366DF
==24056==    at 0xADB2C3: kind (theory_arith.h:409)
==24056==    by 0xADB2C3: get_var_kind (theory_arith.h:543)
==24056==    by 0xADB2C3: is_base (theory_arith.h:544)
==24056==    by 0xADB2C3: smt::theory_arith<smt::mi_ext>::get_a_base_row_that_contains(int) (theory_arith_aux.h:494)
==24056==    by 0xADB3F9: smt::theory_arith<smt::mi_ext>::del_vars(unsigned int) (theory_arith_core.h:3471)
==24056==    by 0xB34CD3: smt::theory_arith<smt::mi_ext>::pop_scope_eh(unsigned int) (theory_arith_core.h:3357)
==24056==    by 0xA1B88A: smt::context::pop_scope_core(unsigned int) (smt_context.cpp:2436)
==24056==    by 0xA1C398: smt::context::pop_scope(unsigned int) (smt_context.cpp:2473)
==24056==    by 0xA4C2F1: ~scoped_ctx_push (smt_model_checker.cpp:291)
==24056==    by 0xA4C2F1: smt::model_checker::check(quantifier*) (smt_model_checker.cpp:299)
==24056==    by 0xA4C491: smt::model_checker::check_quantifiers(bool, bool&, unsigned int&) (smt_model_checker.cpp:487)
==24056==    by 0xA4C80B: smt::model_checker::check(proto_model*, obj_map<smt::enode, app*> const&) (smt_model_checker.cpp:427)
==24056==    by 0xA69237: check_model (smt_quantifier.cpp:702)
==24056==    by 0xA69237: check_model (smt_quantifier.cpp:353)
==24056==    by 0xA69237: smt::quantifier_manager::check_model(proto_model*, obj_map<smt::enode, app*> const&) (smt_quantifier.cpp:461)
==24056==    by 0xA1FE1F: smt::context::restart(lbool&, unsigned int) (smt_context.cpp:3569)
==24056==    by 0xA20548: smt::context::search() (smt_context.cpp:3538)
==24056==    by 0xA20DFA: smt::context::check(unsigned int, expr* const*, bool, bool) (smt_context.cpp:3421)
```

**Comment by NikolajBjorner** [comment]
sure, but this was now fixed.

**Comment by danieldietsch** [comment]
On my end it is fixed. The script does not trigger the bug again. Thank you!

**Comment by invd** [comment]
I agree that the patch likely fixes the issue, but I want to emphasize that since the **unpatched** 4.8.3 did not crash with the provided input, running the **patched* 4.8.3 without issues does not prove as much as we would hope. Practical "stress-testing" with some input variations via Ultimate should confirm this, though. 
In the meantime, I think this issue can be closed. Thanks for your quick work!
